Barotropic: CS mem cleanup; remove btcalc copies

marshallward opened the pull request:
This patch replaces the safer barotropic CS alloc/delete inside of
btcalc and replaces it with a persistent CS created by barotropic_init.

A difficulty here is that btcalc is called by both the dynamic core step
function (e.g. stype_dyn_split_RK2) as well as barotropic_init, meaning
that it needs to be partially set up before btcalc can be called to
finialize the barotropic setup.

This is mostly a matter of timing each event so that the GPU state is
not out of sync with any of the CPU operations.

I'm not going to pretend that I understand everything here.  Multiple
attempts failed, and this is the one that worked.  We still have some
lessons to learn about CS initialization.

marshallward committed:
Barotropic: CS mem cleanup; remove btcalc copies

This patch replaces the safer barotropic CS alloc/delete inside of
btcalc and replaces it with a persistent CS created by barotropic_init.

A difficulty here is that btcalc is called by both the dynamic core step
function (e.g. stype_dyn_split_RK2) as well as barotropic_init, meaning
that it needs to be partially set up before btcalc can be called to
finialize the barotropic setup.

This is mostly a matter of timing each event so that the GPU state is
not out of sync with any of the CPU operations.

I'm not going to pretend that I understand everything here.  Multiple
attempts failed, and this is the one that worked.  We still have some
lessons to learn about CS initialization.

marshallward commented:
@edoyango Can you have a look at this?  This should only allocate the `barotropic_CSp` once, and eliminates the redundant up/downloads in `btcalc`.

The halo transfers are still very extensive, and are likely still the dominant performance concern, but I think this is the right direction.

It seems to pass the "double_gyre/benchmark" test but no doubt I could be more thorough.

marshallward committed:
Barotropic: Minor bt_cont directive cleanup

Rewrote several Bt_cont-related allocation directives:
* Reduce line length to MOM6 style guide (<120)
* Follow ordering in derive type definition
* Use delete: in place of release: for the finalization function

marshallward committed:
Barotropic: btcalc k-loop order fix

Two k-loops were concurrent, but k depends on k+1, so there was a
possibility of error (if not likelihood).

The GPU build may have anticipated this (if not Nvidia entirely) but GNU
seemed to generate NaN errors.  Could this be the cause?

edoyango commented on src/core/MOM_barotropic.F90:
```diff
@@ -6136,12 +6140,9 @@ subroutine barotropic_init(u, v, h, Time, G, GV, US, param_file, diag, CS, &
   !$omp target enter data map(to: CS%D_u_Cor, CS%D_v_Cor)
   !$omp target enter data map(to: CS%dx_Cv, CS%dy_Cu)
-  !$omp target enter data map(to: CS%eta_cor)
-  !$omp target enter data map(to: CS%frhatu, CS%frhatv)
```
For some reason, putting these transfers here give me segfaults in btcalc (nvfortran 24.9). Is it fine to leave these transfers here at the bottom for now? I have no idea why this happens...

marshallward replied:
CS%frhat[uv] need to be allocated before `call btcalc`.  Not entirely surprised there is a problem though, I can't emphasize the headache it was to work through this.  (Not due to the code in any way, more the unintuitiveness of derived type handling).

The `CS%eta_cor` can probably go anywhere.  But odd that it would cause a problem.  I'll try to find an older compiler and test it out.  (Or are we looking at a fixed compiler bug?)

edoyango replied:
Here's the segfault error message:
```
Failing in Thread:1
Accelerator Fatal Error: call to cuStreamSynchronize returned error 700 (CUDA_ERROR_ILLEGAL_ADDRESS): Illegal address during kernel execution
 File: /home/edwardy/MOM6-examples/src/MOM6/src/core/MOM_barotropic.F90
 Function: btcalc:4320
 Line: 4420
```
Running with `nsys nvprof` didn't add any new info.
Running with `compute-sanitizer` (most of the errors were "Invalid __global__ write of size 8 bytes" - showing only the last few errors):
```
========= Invalid __global__ write of size 8 bytes
=========     at mom_barotropic_btcalc_4420+0x8c0 in MOM_barotropic.F90:4424
=========     by thread (28,0,0) in block (0,0,0)
=========     Access at 0xe56ce10 is out of bounds
=========     and is 129,632,140,079,600 bytes before the nearest allocation at 0x75e664000000 of size 67,108,864 bytes
=========     Saved host backtrace up to driver entry point at kernel launch time
=========         Host Frame:  [0x2f285f] in libcuda.so.1
=========         Host Frame: cuda_launch3 in cuda_launch.c:822 [0x1a2b7] in libaccdevice.so
=========         Host Frame: __pgi_uacc_cuda_launch3 in cuda_launch.c:1015 [0x1c509] in libaccdevice.so
=========         Host Frame: __pgi_uacc_cuda_launch in cuda_launche.c:169 [0x1c9ea] in libaccdevice.so
=========         Host Frame: __pgi_uacc_launch in launch.c:54 [0x3b9b4] in libacchost.so
=========         Host Frame: mom_barotropic_btcalc_ in MOM_barotropic.F90:4420 [0x29b857] in MOM6
=========         Host Frame: mom_barotropic_barotropic_init_ in MOM_barotropic.F90:6082 [0x2cf210] in MOM6
=========         Host Frame: mom_dynamics_split_rk2_initialize_dyn_split_rk2_ in MOM_dynamics_split_RK2.F90:1675 [0x466ca7] in MOM6
=========         Host Frame: mom_initialize_mom_ in MOM.F90:3552 [0x70753] in MOM6
=========         Host Frame: MAIN_ in MOM_driver.F90:0 [0x432af0] in MOM6
=========         Host Frame: main [0x97f0] in MOM6
========= 
========= Program hit CUDA_ERROR_LAUNCH_FAILED (error 719) due to "unspecified launch failure" on CUDA API call to cuStreamSynchronize.
=========     Saved host backtrace up to driver entry point at error
=========         Host Frame:  [0x2f19ac] in libcuda.so.1
=========         Host Frame: __pgi_uacc_cuda_wait in cuda_wait.c:82 [0x16c89] in libaccdevice.so
=========         Host Frame: __pgi_uacc_computedone2 in computeexitdone.c:59 [0x2c0c1] in libacchost.so
=========         Host Frame: mom_barotropic_btcalc_ in MOM_barotropic.F90:4434 [0x29aa8b] in MOM6
=========         Host Frame: mom_barotropic_barotropic_init_ in MOM_barotropic.F90:6082 [0x2cf210] in MOM6
=========         Host Frame: mom_dynamics_split_rk2_initialize_dyn_split_rk2_ in MOM_dynamics_split_RK2.F90:1675 [0x466ca7] in MOM6
=========         Host Frame: mom_initialize_mom_ in MOM.F90:3552 [0x70753] in MOM6
=========         Host Frame: MAIN_ in MOM_driver.F90:0 [0x432af0] in MOM6
=========         Host Frame: main [0x97f0] in MOM6
========= 
Failing in Thread:1
========= Program hit CUDA_ERROR_LAUNCH_FAILED (error 719) due to "unspecified launch failure" on CUDA API call to cuCtxSynchronize.
=========     Saved host backtrace up to driver entry point at error
=========         Host Frame:  [0x2887ba] in libcuda.so.1
=========         Host Frame: __pgi_uacc_cuda_error_handler in cuda_error.c:101 [0xfb93] in libaccdevice.so
=========         Host Frame: __pgi_uacc_cuda_wait in cuda_wait.c:84 [0x16ef5] in libaccdevice.so
=========         Host Frame: __pgi_uacc_computedone2 in computeexitdone.c:59 [0x2c0c1] in libacchost.so
=========         Host Frame: mom_barotropic_btcalc_ in MOM_barotropic.F90:4434 [0x29aa8b] in MOM6
=========         Host Frame: mom_barotropic_barotropic_init_ in MOM_barotropic.F90:6082 [0x2cf210] in MOM6
=========         Host Frame: mom_dynamics_split_rk2_initialize_dyn_split_rk2_ in MOM_dynamics_split_RK2.F90:1675 [0x466ca7] in MOM6
=========         Host Frame: mom_initialize_mom_ in MOM.F90:3552 [0x70753] in MOM6
=========         Host Frame: MAIN_ in MOM_driver.F90:0 [0x432af0] in MOM6
=========         Host Frame: main [0x97f0] in MOM6
========= 
Accelerator Fatal Error: call to cuStreamSynchronize returned error 719 (CUDA_ERROR_LAUNCH_FAILED): Launch failed (often invalid pointer dereference)
 File: /home/edwardy/MOM6-examples/src/MOM6/src/core/MOM_barotropic.F90
 Function: btcalc:4320
 Line: 4420

========= Target application returned an error
========= ERROR SUMMARY: 41 errors
```

edoyango replied:
I noticed that removing `CS` from the update in line 6081 causes the segfaults to disappear. Naturally, I get wrong results since lots of stuff isn't updated.

Perhaps having `!$omp target enter data map(to: CS, CS%fhratu, CS%fhratv, CS%eta_cor)` just after reading in all the params and removing the `alloc: CS%barotropic_CSp` outside the subroutine can achieve the goal of saving me from segfaults and removing data transfers from `btcalc`?

marshallward replied:
`compute-sanitizer` wasn't able to replicate your error, but I did find a different one:
```
Failing in Thread:1
Accelerator Fatal Error: call to cuStreamSynchronize returned error 719 (CUDA_ERROR_LAUNCH_FAILED): Launch failed (often invalid pointer dereference)
 File: /home/Marshall.Ward/gfdl/m6e/src/MOM6/src/core/MOM_barotropic.F90
 Function: set_local_bt_cont_types:4864
 Line: 4946
```
My error summary is also far worse :grimacing: :
```
========= Target application returned an error
========= ERROR SUMMARY: 2562 errors
========= ERROR SUMMARY: 2462 errors were not printed. Use --print-limit option to adjust the number of printed errors
```
Also, perhaps this is relevant, but I noticed a few of the nodes on our machine would just fail and repeatedly spam `Failing in Thread:1`.  I had thought it was just a bad node, but perhaps this was just hitting these underlying errors.

Looks like I have some work to do...

marshallward replied:
My issue appears to be due to `call swap` (subroutines strike again!).  But I'll move it to a separate comment if I have anything more to add.

The biggest lesson for me here is that `compute-sanitizer` should probably be part of our acceptance testing.

marshallward replied:
I tried replacing the external `map(alloc: CS)` with an internal `map(to: CS)` and pushing it as close to the first internal `call btcalc` as possible, but it seems to cause the `benchmark` config to break.  I struggle to explain this.  The "distance" between the `map()` and the `update()` is only a few seemingly inconsequential lines.

```
   if (CS%id_frhatu1 > 0) allocate(CS%frhatu1(IsdB:IedB,jsd:jed,nz), source=0.)
   if (CS%id_frhatv1 > 0) allocate(CS%frhatv1(isd:ied,JsdB:JedB,nz), source=0.)
 
+  !$omp target enter data map(to: CS)
+  !$omp target enter data map(to: CS%frhatu, CS%frhatv)
+  !$omp target enter data map(to: CS%eta_cor)
+
   if (.NOT.query_initialized(CS%ubtav,"ubtav",restart_CS) .or. &
       .NOT.query_initialized(CS%vbtav,"vbtav",restart_CS)) then
+    !!$omp target update to(h)
+    !$omp target update to(CS,h)
     call btcalc(h, G, GV, CS, may_use_default=.true.)
     CS%ubtav(:,:) = 0.0 ; CS%vbtav(:,:) = 0.0
     do k=1,nz ; do j=js,je ; do I=is-1,ie
```
The `update(CS, H)` works, the `update(h)` does not.

marshallward replied:
I tested this on another system with v24.5 and can now replicate the error:
```
Failing in Thread:1
Accelerator Fatal Error: call to cuStreamSynchronize returned error 700 (CUDA_ERROR_ILLEGAL_ADDRESS): Illegal address during kernel execution
 File: $M6E/src/MOM6/src/core/MOM_barotropic.F90
 Function: btcalc:4320
 Line: 4420
```
We could call this "compiler bug", but I'll see if there's some way to get around it

marshallward replied:
I updated the branch to follow your suggestion of allocate-and-copying `CS` inside of `barotropic_init`.  It was failing `benchmark`, but this was because there happened to be a CPU usage of `CS%frhat[uv]` in `set_dtbt()`.

Perhaps `set_dtbt`'s work could be moved to GPU, but for now I added an `update from()`.

No idea why it was only a problem after the `CS` allocation change, but the `set_dtbt` update fixed `benchmark` while allowing for the internal `CS` allocate.

I was able to test this to some limited degree.  I'll save the details for our next meeting, but it does seem to work on v24.5.

edoyango replied:
Your changes seem to have done the trick - working for both double_gyre and benchmark.

edoyango commented on src/core/MOM_dynamics_split_RK2.F90:
```diff
@@ -662,6 +664,7 @@ subroutine step_MOM_dyn_split_RK2(u_inst, v_inst, h, tv, visc, Time_local, dt, f
     !$omp target update to(CS%BT_cont%h_u, CS%BT_cont%h_v)
     call cpu_clock_end(id_clock_continuity)
     if (BT_cont_BT_thick) then
+      !$omp target update to(h, CS%BT_cont%h_u, CS%BT_cont%h_v)
```
There's a `bt_mass_source` below that uses `hp`, which hasn't yet been mapped to GPU. That was the only implicit transfer of data I found, which you probably want to get rid of (or make explicit).

marshallward replied:
This should be added now (just a simple move of upload from bt_mass_source() to dyncore loop).

edoyango reviewed (changes requested):
Besides the issue of when to transfer `CS%{eta_cor,frhatu,frhatv}` in `barotropic_init`, this works on my laptop in the `benchmark` and `double_gyre` cases! Using `NV_ACC_NOTIFY`, I didn't see many implicit transfers and your additional updates seem to make sense.

marshallward commented:
The errors detected by `compute-sanitizer` mentioned above were from `call swap()` inside of `set_local_bt_cont_types`.  In this case, they can be averted by just not using the function.

I tried to somehow get this working with `!$omp declare target` but no luck.  This probably needs more analysis, and perhaps reported to Nvidia.

~Most unfortunately, I still can't replicate the error seen by @edoyango.  Perhaps its a compiler-specific issue (24.11 vs ?).  I still have an A100 with 25.5 that I can check.~

I think I can replicate it now, see the comments above.

marshallward committed:
Barotropic: No swap() with derived types

Calls to swap() inside of set_local_bt_cont_types have been removed and
replaced with explicit swaps.  Although these functions could be called
and used inside do concurrent GPU kernels, the compute-sanitizer was
detecting address errors, which might indicate an underlying problem, so
they have been removed.

Another instance of call swap() was kept, since no error has yet been
detected.  (I don't know if this is actually called, but it does not use
derived types so there may be no issue.)

Two potentially redundant BT_cont%[uv] uploads after continuity calls
have been removed.  We can deal with this after the continuity solvers
have been ported to GPU.

marshallward committed:
Barotropic: Move hp uploads to dyncore loop

edoyango commented:
> Most unfortunately, I still can't replicate the error seen by @edoyango. Perhaps its a compiler-specific issue (24.11 vs ?). I still have an A100 with 25.5 that I can check.

[Jorge and I posted on the nvidia dev forums](https://forums.developer.nvidia.com/t/nvfortran-openmp-offload-segfaults-when-updating-derived-type-with-allocatable-member/340161) and we're pretty sure the `update` is buggy for derived types with allocatable members. `update to(CS)` copies over the host address of the member allocatables and overwrites the device address, so when an offloaded loop tries to access the allocatable error, it segfaults. But this only happens with `-gpu=mem:separate`. Are you running on the GH chips? I think they have the same address space, so it's using the correct address? idk.

It implies we can't safely use `!$omp target update` with derived types that have allocatable members

marshallward committed:
Barotropic: Move CS alloc inside of init

The `!$omp target update to(CS)` inside of `barotropic_init` seemed
choke on some Nvidia compiler versions.  It was perhaps unnecessary,
since we could replace it will a well-timed `map(to: CS)` inside of
`barotropic_init`.

Naively swapping did not work because of a missing use of CS%frhat[uv]
happening on CPU-side.  (Perhaps this should not be, but for now it is
the case...)

By making both change, the `benchmark` case is again working.
